store/sqlstore/channel_store.go

```diff
@@ -2350,6 +2346,7 @@ func (s SqlChannelStore) ClearAllCustomRoleAssignments() store.StoreChannel {
 				result.Err = model.NewAppError("SqlChannelStore.ClearAllCustomRoleAssignments", "store.sql_channel.clear_all_custom_role_assignments.open_transaction.app_error", nil, err.Error(), http.StatusInternalServerError)
 				return
 			}
+			defer transaction.Rollback()
 
 			var channelMembers []*channelMember
 			if _, err := transaction.Select(&channelMembers, "SELECT * from ChannelMembers WHERE (ChannelId, UserId) > (:ChannelId, :UserId) ORDER BY ChannelId, UserId LIMIT 1000", map[string]interface{}{"ChannelId": lastChannelId, "UserId": lastUserId}); err != nil {
```

Comment:
On the line below, we check `rollback` and return an error, which makes it a bit harder to understand the control flow given the new `defer`.

I wonder if we should create a helper method:
```go
// tearDownTransaction ensures a transaction is closed after use
tearDownTransaction(transaction *gorp.Transaction) {
    // Rollback is a no-op if the transaction was committed.
    if err := transaction.Rollback(); err != nil {
        mlog.Error("Failed to rollback transaction", mlog.Err(err))
    }
}
```

and then use this in place of the direct `rollback`:
```
if transaction, err = s.GetMaster().Begin(); err != nil {
    // ...
}
defer tearDownTransaction(transaction)
```

Then we can remove these extra rollback checks altogether.

Of course, this means we're no longer returning an error on a failed rollback, but I think that's reasonable, since we probably want to return the original error that caused us to want to rollback in the first place.

Reply:
Mmh, I'm not sure this is the best way to go, here. First, because `defer`'d calls to `tearDownTransaction` would report errors even for transactions which were successfully committed. This would unnecessarily pollute the error logs.

Also, when looking at the code, and how `Commit()` and `Rollback()` are implemented in `gorp`, it seems that these pieces of code in Mattermost were slightly buggy to begin with. Indeed, when a `Commit()` fails, calling `Rollback()` afterwards will **always** fail, as the first thing `Commit()` does is to set the transaction's `closed` attribute to `true` (see [l.151](https://github.com/go-gorp/gorp/blob/f3677d4a0a8838c846ed41bf41927f2c8713bd60/transaction.go#L151) of `gorp/transaction.go`), which then forces all subsequent calls to `Rollback()` to immediately fail with `sql.ErrTxDone` (see [l.163-174](https://github.com/go-gorp/gorp/blob/f3677d4a0a8838c846ed41bf41927f2c8713bd60/transaction.go#L164-L173)).

Therefore, calling `Rollback()` right after a `Commit()` failed is not such a good idea to begin with: the `Rollback()` will systematically fail with `sql.ErrTxDone`, and we will lose the ability to report the probably more informative `Commit()` error message. I would therefore suggest to just remove the explicit calls to `Rollback()` you just mentioned, and report the `Commit()` error instead in case of failure.

What do you think?

Reply:
Excellent analysis! This is further clarified by `database/sql` [itself](https://golang.org/pkg/database/sql/#Tx):
> Tx is an in-progress database transaction.
A transaction must end with a call to Commit or Rollback.
After a call to Commit or Rollback, all operations on the transaction fail with ErrTxDone.
The statements prepared for a transaction by calling the transaction's Prepare or Stmt methods are closed by the call to Commit or Rollback.

So, indeed, any `Rollback` after an attempt to `Commit` can be safely removed. Let's do it!

Reply:
As a further note, would it still make sense to introduce a:
```go
// closeTransaction ensures a transaction is closed after use, rolling back if not committed
closeTransaction(transaction *gorp.Transaction) {
    // Rollback returns sql.ErrTxDone if the transaction was already closed
    if err := transaction.Rollback(); err != nil && err != sql.ErrTxDone {
        mlog.Error("Failed to rollback transaction", mlog.Err(err))
    }
}
```

just to clarify what's going on in the code?

Reply:
1/5 would prefer:

```go
// finalizeTransaction, handles committing or rolling back a transaction, depending on a priorError value 
func finalizeTransaction(transaction *gorp.Transaction, priorError error) error {
    if priorError ==  nil {
        return transaction.Commit()
    }
    // Rollback returns sql.ErrTxDone if the transaction was already closed
    err := transaction.Rollback()
    if err != nil && err != sql.ErrTxDone {
        mlog.Error("Failed to rollback transaction", mlog.Err(err))
    }
    return priorError
}

func SaveSomething() (..., returnError error){
    ....
    transaction, err := s.GetMaster().Begin()
    if err != nil {
         return err
    }
    defer func () {
        returnError = finalizeTransaction(transaction, returnError)
    }()
    ...
}
```

I understand that ^^ is an overgeneralization, but then exceptions can just be coded explicitly.

Reply:
I definitely like the nomenclature, `finalizeTransaction`. I confess I'm not a big fan of named return parameters (mostly because they enable naked returns, which can in turn mask issues), but in practice this won't be necessary, since the actual code in these stores probably looks more like:
```
   defer func () {
        result.Err = finalizeTransaction(transaction, result.Err)
    }()
```

My only gripe with the above is that the `Commit` (or `Rollback`, if done programmatically) is no longer explicit. I personally like deferred functions for cleanup, but it feels awkward to rely on them for the "happy path". I'm only 1/5 on this myself though, since I think this might only be personal preference.

Reply:
I am ok with leaving this to the committer, either way is really ok with me. I support the idea of a deferrable convenience method.